### ISSUE TITLE
[typedef] Option proposal: allow inferred types

### ISSUE THREAD

**Issue by MatiasProietti** [typescript-eslint/typescript-eslint#4276]
### Description

I couldn't decide between this template or the rule issue one. This one seems more appropriate considering this is a proposal to add an option on an existing rule,

Both inferred types and the typedef rule help maintain a clean and readable code, but sadly right now you are forced to declare the type even when it can be inferred.

If we have this option set to true then we can use `typedef` and `no-inferrable-types` together.

### Fail

When the option `"inferredTypes": false`:
```ts
title = 'yet-another-angular-boilerplate';
const = 5;
```

### Pass

When the option `"inferredTypes": true`:
```ts
title = 'yet-another-angular-boilerplate';
const = 5;
```

When the option `"inferredTypes": false`:
```ts
title: string = 'yet-another-angular-boilerplate';
const: number = 5;
```

**Comment by bradzacher** [comment]
`typedef` is intended to be a transitional rule. It's there to help you find places that could be annotated to onboard a codebase to compiler options like `--noImplicitAny` and `--strictPropertyInitialization`.

As such, we are not accepting options which make it easier to use forever on a codebase.

**Comment by MatiasProietti** [comment]
Thank you, I'm still in the learning process so this answer helps a lot.
I know this is off topic but do you recommend any book that I can read on the subject? I would like to learn more about this.

**Comment by bradzacher** [comment]
I don't have any books sorry as I'm personally someone who learns by doing.

I know there are plenty of books available at all levels of TS knowledge and there is no shortage of blogs on the subject as well.

As for the nitty gritty like perf - that is definitely just learning by doing - using the compiler APIs and such - and by talking with the TS team. There's not really any books on the subject there.
